File: lexer/lexer.go
package lexer

import (
	"zunino.br/monkey-lang/token"
)

func New(input string) *Lexer {
	lexer := &Lexer{input: input}
	lexer.readChar()
	return lexer
}

type Lexer struct {
	input        string
	position     int
	readPosition int
	ch           byte
}

func newToken(tokType token.TokenType, ch byte) token.Token {
	return token.Token{Type: tokType, Text: string(ch)}
}

func (l *Lexer) NextToken() token.Token {
	var tok token.Token

	l.skipWhitespace()

	switch l.ch {
	case '=':
        if l.peekChar() == '=' {
            l.readChar()
            tok = token.Token{Type: token.EQUALS, Text: "=="}
        } else {
            tok = newToken(token.ASSIGN, l.ch)
        }
	case ',':
		tok = newToken(token.COMMA, l.ch)
	case ';':
		tok = newToken(token.SEMICOLON, l.ch)
	case '(':
		tok = newToken(token.LPAREN, l.ch)
	case ')':
		tok = newToken(token.RPAREN, l.ch)
	case '{':
		tok = newToken(token.LBRACE, l.ch)
	case '}':
		tok = newToken(token.RBRACE, l.ch)
	case '+':
		tok = newToken(token.PLUS, l.ch)
    case '-':
        tok = newToken(token.MINUS, l.ch)
    case '/':
        tok = newToken(token.SLASH, l.ch)
    case '*':
        tok = newToken(token.ASTERISK, l.ch)
    case '<':
        tok = newToken(token.LT, l.ch)
    case '>':
        tok = newToken(token.GT, l.ch)
    case '!':
        if l.peekChar() == '=' {
            l.readChar()
            tok = token.Token{Type: token.DIFFERENT, Text: "!="}
        } else {
            tok = newToken(token.BANG, l.ch)
        }
	case 0:
		tok.Text = ""
		tok.Type = token.EOF
	default:
		if isLetter(l.ch) {
			tok.Text = l.readIdentifier()
			tok.Type = token.LookupIdent(tok.Text)
			return tok
		} else if isDigit(l.ch) {
			tok.Text = l.readInt()
			tok.Type = token.INT
			return tok
		}
		tok = newToken(token.ILLEGAL, l.ch)
	}

	l.readChar()
	return tok
}

func (l *Lexer) readChar() {
	if l.readPosition >= len(l.input) {
		l.ch = 0
	} else {
		l.ch = l.input[l.readPosition]
	}
	l.position = l.readPosition
	l.readPosition++
}

func (l *Lexer) readIdentifier() string {
	position := l.position
	for isLetter(l.ch) {
		l.readChar()
	}
	return l.input[position:l.position]
}

func (l *Lexer) readInt() string {
	position := l.position
	for isDigit(l.ch) {
		l.readChar()
	}
	return l.input[position:l.position]
}

func (l *Lexer) skipWhitespace() {
	for l.ch == ' ' || l.ch == '\t' || l.ch == '\n' || l.ch == '\r' {
		l.readChar()
	}
}

func (l *Lexer) peekChar() byte {
    if l.readPosition <  len(l.input) {
        return l.input[l.readPosition]
    }
    return 0
}

func isLetter(ch byte) bool {
	return 'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z' || ch == '_'
}

func isDigit(ch byte) bool {
	return ch >= '0' && ch <= '9'
}

File: rust/src/lexer.rs
use crate::token::{get_keyword, Token};
use std::str;

pub struct Lexer {
    input: Vec<u8>,
    pos: usize,
    rpos: usize,
    ch: u8,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        let mut lexer = Self {
            input: input.bytes().collect(),
            pos: 0,
            rpos: 0,
            ch: 0,
        };
        lexer.consume_char();
        lexer
    }

    pub fn next_token(&mut self) -> Token {
        self.skip_whitespace();
        let token = match self.ch {
            b'=' => Token::Assign,
            b',' => Token::Comma,
            b';' => Token::Semicolon,
            b'+' => Token::Plus,
            b'(' => Token::Lparen,
            b')' => Token::Rparen,
            b'{' => Token::Lbrace,
            b'}' => Token::Rbrace,
            _ if self.ch.is_ascii_digit() => self.read_number(),
            _ if self.ch.is_ascii_alphabetic() => self.read_word(),
            0u8 => Token::Eof,
            _ => Token::Illegal,
        };
        self.consume_char();
        token
    }

    fn consume_char(&mut self) {
        if self.rpos >= self.input.len() {
            self.ch = 0;
        } else {
            self.ch = self.input[self.rpos];
        }
        self.pos = self.rpos;
        self.rpos += 1;
    }

    fn skip_whitespace(&mut self) {
        while self.ch.is_ascii_whitespace() {
            self.consume_char();
        }
    }

    fn rewind(&mut self) {
        self.pos -= 1;
        self.rpos = self.pos + 1;
    }

    fn read_word(&mut self) -> Token {
        let position = self.pos;

        while self.ch.is_ascii_alphabetic() {
            self.consume_char();
        }

        let word = str::from_utf8(&self.input[position..self.pos])
            .unwrap()
            .to_owned();

        self.rewind();

        if let Some(keyword_token) = get_keyword(&word) {
            return keyword_token.clone();
        }

        Token::Ident(word.to_string())
    }

    fn read_number(&mut self) -> Token {
        let position = self.pos;

        while self.ch.is_ascii_digit() {
            self.consume_char();
        }

        let number = str::from_utf8(&self.input[position..self.pos])
            .unwrap()
            .parse::<i64>()
            .unwrap();

        self.rewind();

        Token::Int(number)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_next_token() {
        let input = r"let   five  =   5;
let ten = 10;

let add = fn(x, y) {
    x + y;
};

let result = add(five, ten);";
        let mut lexer = Lexer::new(input);

        let expected = vec![
            Token::Let,
            Token::Ident("five".to_string()),
            Token::Assign,
            Token::Int(5),
            Token::Semicolon,
            Token::Let,
            Token::Ident("ten".to_string()),
            Token::Assign,
            Token::Int(10),
            Token::Semicolon,
            Token::Let,
            Token::Ident("add".to_string()),
            Token::Assign,
            Token::Function,
            Token::Lparen,
            Token::Ident("x".to_string()),
            Token::Comma,
            Token::Ident("y".to_string()),
            Token::Rparen,
            Token::Lbrace,
            Token::Ident("x".to_string()),
            Token::Plus,
            Token::Ident("y".to_string()),
            Token::Semicolon,
            Token::Rbrace,
            Token::Semicolon,
            Token::Let,
            Token::Ident("result".to_string()),
            Token::Assign,
            Token::Ident("add".to_string()),
            Token::Lparen,
            Token::Ident("five".to_string()),
            Token::Comma,
            Token::Ident("ten".to_string()),
            Token::Rparen,
            Token::Semicolon,
            Token::Eof,
        ];

        for t in 0..expected.len() {
            let tok = lexer.next_token();
            assert_eq!(tok, expected[t]);
        }
    }

    #[test]
    fn test_read_char_should_read_all_chars_from_given_input() {
        let input = r"BAR";
        let mut lexer = Lexer::new(input);
        assert_eq!(lexer.ch, 'B' as u8);
        lexer.consume_char();
        assert_eq!(lexer.ch, 'A' as u8);
        lexer.consume_char();
        assert_eq!(lexer.ch, 'R' as u8);
        lexer.consume_char();
        assert_eq!(lexer.ch, 0 as u8);
    }
}
